packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -192,74 +183,124 @@ export class MetadataEditor extends ReactWidget {
 
-  onAfterShow(): void {
-    if (!this.editor) {
+  onUpdateRequest(msg: Message): void {
+    super.onUpdateRequest(msg);
+    if (!this.editor && this.metadata['code']) {
       const getMimeTypeByLanguage = this.editorServices.mimeTypeService
         .getMimeTypeByLanguage;
       this.editor = this.editorServices.factoryService.newInlineEditor({
-        host: document.getElementById('Code:' + this.name),
+        host: document.getElementById('code:' + this.id),
         model: new CodeEditor.Model({
-          value: this.getFormItem('Code').value,
+          value: this.metadata['code'].join('\n'),
           mimeType: getMimeTypeByLanguage({
-            name: this.getFormItem('Language').value.choice,
-            codemirror_mode: this.getFormItem('Language').value.choice
+            name: this.metadata['language'],
+            codemirror_mode: this.metadata['language']
           })
         })
       });
       this.editor.model.value.changed.connect((args: any) => {
-        this.getFormItem('Code').value = args.text;
+        this.metadata['code'] = args.text.split('\n');
         this.handleDirtyState(true);
       });
     }
   }
 
+  getDefaultChoices(fieldName: string) {
+    let defaultChoices = this.schema[fieldName].uihints.default_choices;
+    if (defaultChoices == undefined) {
+      defaultChoices = [];
+    }
+    for (const otherMetadata of this.allMetadata) {
+      if (!defaultChoices.includes(otherMetadata.metadata[fieldName])) {
+        defaultChoices.push(otherMetadata.metadata[fieldName]);
+      }
+    }
+    return defaultChoices;
+  }
+
+  renderTextInput(
+    label: string,
+    description: string,
+    fieldName: string,
+    defaultValue
+  ) {
+    return (
+      <FormGroup
+        key={label}
+        label={label}
+        labelInfo="(required)"
+        helperText={description}
+      >
+        <InputGroup
+          onChange={(event: any): void => {
+            this.handleTextInputChange(event, fieldName);
+          }}
+          defaultValue={defaultValue}
+          type="text-input"
+        />
+      </FormGroup>
+    );
+  }
+
+  renderField(fieldName: string) {
+    const uihints = this.schema[fieldName].uihints;
+    if (uihints == undefined) {
+      return;
+    } else if (
+      uihints.field_type == 'textinput' ||
+      uihints.field_type == undefined
+    ) {
+      return this.renderTextInput(
+        uihints.label,
+        uihints.description,
+        fieldName,
+        this.metadata[fieldName]
+      );
+    } else if (uihints.field_type == 'dropdown') {
+      return (
+        <DropDown
+          label={uihints.label}
+          schemaField={fieldName}
+          description={uihints.description}
+          choice={this.metadata[fieldName]}
+          defaultChoices={this.getDefaultChoices(fieldName)}
+          handleDropdownChange={this.handleDropdownChange}
+        ></DropDown>
+      );
+    } else if (uihints.field_type == 'code') {
+      return (
+        <>
```

Comment:
is this valid dom? shouldn't this need to be a div or span or something?

Reply:
It doesn't seem to cause any issues, but I'll add `div` because it seems more reliable